### PR TITLE
lightning network common misconceptions corrected

### PR DESCRIPTION
**Pull request by zack-bitcoin**
GNUclear is looking good.
I still think that it would work better if you added lightning consensus optimization.
Tendermint consensus and channel consensus aren't contradictory protocols, they can work together beautifully. It would mix the bonding mechanism with the channels mechanism.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1241,23 +1244,28 @@ and the Dfinity project. [\[18\]][18]
...
-mechanism, for cost-saving and privacy reasons.
+Network can become focal points for routing the payments, and contracts of others, leading to a
+fully connected channel network, at the cost of capital being tied up on
+payment channels. This cost can be completely removed by making it into
+a useful proof of work. Like the [lightning consensus protocol](https://github.com/BumblebeeBat/FlyingFox/blob/master/docs/lightning_consensus.md)
+
+The Lightning Network can also easily extend across multiple independent
+blockchains to allow for the transfer of value via an exchange market, _and_
+it's state-channels 
+can be used to assymetrically transfer _tokens_ from one blockchain to
```
Comment (jaekwon) [inline README.md]: What?  I think we disagree about what asymmetric token transfer means.
You can have asymmetric token transfers in GnuClear because there's a central point of record -- the hub.  What is the authority of token ownership in a cross-chain state channel?

File: README.md [hunk 2]
```diff
@@ -1241,23 +1244,28 @@ and the Dfinity project. [\[18\]][18]
...
-While the Lightning Network can also easily extend across multiple independent
-blockchains to allow for the transfer of _value_ via an exchange market, it
-cannot be used to assymetrically transfer _tokens_ from one blockchain to
-another.  The main benefit of the GnuClear network described here is to enable
-such direct token transfers.  That said, we expect payment channels and the
-Lightning Network to become widely adopted along with our token transfer
-mechanism, for cost-saving and privacy reasons.
+Network can become focal points for routing the payments, and contracts of others, leading to a
+fully connected channel network, at the cost of capital being tied up on
+payment channels. This cost can be completely removed by making it into
```
Comment (jaekwon) [inline README.md]: From the link,
> The value that gets destroyed for maintaining channels can simultaneously be used to secure blockchain consensus, so we don't have to destroy value twice. We can kill 2 birds with 1 stone.

So it's not really removing the cost, it's merging two costs.  But I'm not sure that holding gnuts incur any such interest costs, it's not meant to be a medium of exchange.

File: README.md [hunk 3]
```diff
@@ -89,7 +89,10 @@ limited by the capacity of a single physical machine, lest we sacrifice the
 property of complete auditability.  The Lightning Network [\[7\]][7] can help
 scale Bitcoin transaction volume by leaving some transactions off the ledger
 completely and is well suited for micropayments and privacy preserving payment
-rails, but may not be suitable for more generalized scaling needs.
+rails, and it is suitable for more generalized scaling needs.
+<!---
+state-channels allow for generalized smart contracts in the lightning network
+--->
```
Comment (jaekwon) [inline README.md]: Yes but ultimately it's about tokens, yes?  Like, a state-channel is an impermanent thing, and the difference between before-and-after a state channel is that some tokens had moved.  So ultimately it's only about token transfers, within the same ledger (though it can happen atomically with such transfers in one or more other ledgers, it's still not what we call "asymmetric token transfer")

File: README.md [hunk 4]
```diff
@@ -1241,23 +1244,28 @@ and the Dfinity project. [\[18\]][18]
 The Lightning Network is a proposed message relay network operating at a layer
 above the Bitcoin blockchain, enabling many orders of magnitude improvement in
 transaction throughput by moving the majority of transactions outside of the
-consensus ledger into so-called "payment channels". This is made possible (with
-great difficulty) by the Bitcoin scripting language, which enables parties to
+consensus ledger into so-called "payment channels".
+<!---
+lightning network exists beyond bitcoin
+--->
```
Comment (jaekwon) [inline README.md]: agreed, thanks!

File: README.md (hunk 3 above)
Comment (zack-bitcoin) [inline README.md]: I don't understand your comment. What is about tokens? why does impermanence matter? why do you say "asymmetric token transfer" and what does it mean?

Channels have generalized smart contracts the same as on-chain contracts.
Domain registration contracts are less efficient off-chain, but most other things are more efficient off-chain.

The lightning network is not constrained to a single ledger. You can have smart contracts that span multiple blockchains.

File: README.md (hunk 2 above)
Comment (zack-bitcoin) [inline README.md]: The reserve pool is losing money by the interest rate too.

before you said "I see no problem with rewarding the gnut holders"
And now you are saying that there is no extra cost. 
This is contradictory. You can't have it both ways.
You can't pay extra money to gnut holders unless you take extra money from the users.

Flying Fox is set up so that if validators earn too much money, then the reward for making a channel gets higher, and the cost to be a validator gets higher. So the extra profit goes back to the users, making the cost of using the blockchain lower.

File: README.md (hunk 1 above)
Comment (zack-bitcoin) [inline README.md]: is "asymmetric token transfer" defined somewhere? I know what those words are each on their own.

Channel contracts are turing complete, so you can write any program that you can make with smart contracts.
Financial derivatives are key. You can hold any asset so long as your partner is willing to hold the inverse.

File: README.md (hunk 3 above)
Comment (jaekwon) [inline README.md]: Is a state channel a bilateral contract between two users of a ledger?

> Domain registration contracts are less efficient off-chain
How would this even work?  Why would a name-reg system go in a channel?

> The lightning network is not constrained to a single ledger.
I'm not saying otherwise.

File: README.md (hunk 2 above)
Comment (jaekwon) [inline README.md]: > The reserve pool is losing money by the interest rate too.

Yes, I was referring to gnuts without the reserve-pool.

But, I'm not yet convinced that this "interest cost" is real, or that the money paid to gnut holders is "extra/too-much money" that ought to be redistributed to users who paid for them.  My goal is not to make the cost of using the blockchain lower, I'm interested in making a useful system that people will gladly pay to use.

File: README.md (hunk 3 above)
Comment (zack-bitcoin) [inline README.md]: you could use the lightning state network to send an asset linked to the price of gold from one chain to another. bitcoin and litecoin don't off state-channels, they only have payment channels.

name-reg usually wouldn't go into channels, although it could.
If we are playing a game of poker, there could be 5 players who are matched to the 5 names "player 1", "player 2" ... "player 5".  that sort of name-reg would make sense in channels.
But domain names are much less likely, but technically we could do them too.
state-channels are a turing complete platform upon which we can implement consensus mechanisms like tendermint or flying fox or POW bitcoin. 
They are like sharding in that they are a scalability tool.

For domain registration: each participant makes a channel with a central node.
All the channels simultaneously load the same state which links a pubkey to a username for each person.
The channel can be updated with transactions that spend tokens from one user to another by username, you don't have to write the pubkey in the transaction, since the pubkey is written in the channel state.

Comment (jaekwon) [inline README.md]: Ok, before we get into any further detail, what does this have to do with the Lightning Network?

https://lightning.network/

> you could use the lightning state network to send an asset linked to the price of gold from one chain to another. bitcoin and litecoin don't off state-channels, they only have payment channels.

Where is the canonical record of who owns how many of this asset, and what keeps the total amount of this asset from inflating beyond what it should?

File: README.md (hunk 2 above)
Comment (zack-bitcoin) [inline README.md]: I said some confusing things. sorry.

The interest cost is real. 
How much would you pay in bitcoin right now for 10 bitcoin right now?
How much would you pay in bitcoin right now for 10 bitcoin in a month?
The difference between those 2 numbers is the interest cost for a month.

The value that is lost by the interest rate is destroyed. It cannot be given to gnut holders as profit. Even if you want to give an extra reward to gnut holders, you still shouldn't make their bonds do 2 things simultaneous. The security deposit against double signing should not also be used to decide who has power in consensus.

I think it is more complex to give consensus power to $10 of bonds to secure $100 of money. That is as hard as balancing a mattress on a book. 
It is simpler to let the owner of the $100 of money have the consensus power over their own funds. Like balancing a mattress on a mattress. 

I have never suggested that validators should be able to spend or use bonded tokens. That is the exact opposite of what I am saying.
I am suggesting that we use the bonded coins only as a security deposit against double-signing, and to not re-use it for choosing who has power in consensus.

>If we can find some use to those bonded reserve-pool collateral tokens, we'll implement it.
That is exactly what I am telling you not to do. The bonds should be used for security deposit against double-signing, and nothing else. 

>But I'm not convinced that channels are the solution
I am not telling you to use channels. That is one way to solve this problem, but it is not the only way to solve this problem.
I am telling you that you can't re-use bonds for multiple things.
DPOS could work too. Or rewarding everyone with their computer turned on, and having every node participate. 

If you use the same bonds to both secure against double-signing and to decide who has power in consensus, it will do both things poorly.
The size of bonds will be too big to optimally secure against double-signing which will waste money, and it will be too small to decide who should have power in consensus which will cause retirement attacks from validators.

Comment (zack-bitcoin) [inline README.md]: I said some confusing things. sorry.

The interest cost is real. 
How much would you pay in bitcoin right now for 10 bitcoin right now?
How much would you pay in bitcoin right now for 10 bitcoin in a month?
The difference between those 2 numbers is the interest cost for a month.

The value that is lost by the interest rate is destroyed. It cannot be given to gnut holders as profit. Even if you want to give an extra reward to gnut holders, you still shouldn't make their bonds do 2 things simultaneous. The security deposit against double signing should not also be used to decide who has power in consensus.

I think it is more complex to give consensus power to $10 of bonds to secure $100 of money. That is as hard as balancing a mattress on a book. 
It is simpler to let the owner of the $100 of money have the consensus power over their own funds. Like balancing a mattress on a mattress. 

I have never suggested that validators should be able to spend or use bonded tokens. That is the exact opposite of what I am saying.
I am suggesting that we use the bonded coins only as a security deposit against double-signing, and to not re-use it for choosing who has power in consensus.

>If we can find some use to those bonded reserve-pool collateral tokens, we'll implement it.
That is exactly what I am telling you not to do. The bonds should be used for security deposit against double-signing, and nothing else. 

>But I'm not convinced that channels are the solution
I am not telling you to use channels. That is one way to solve this problem, but it is not the only way to solve this problem.
I am telling you that you can't re-use bonds for multiple things.
DPOS could work too. Or rewarding everyone with their computer turned on, and having every node participate. 

If you use the same bonds to both secure against double-signing and to decide who has power in consensus, it will do both things poorly.
The size of bonds will be too big to optimally secure against double-signing which will waste money, and it will be too small to decide who should have power in consensus which will cause retirement attacks from validators.

File: README.md (hunk 3 above)
Comment (zack-bitcoin) [inline README.md]: If you and I have a channel with $1000 in it, then we couldn't have more than $1000 of anything in our channel.
We could make it have $100 linked to the price of gold, and $100 that moves inverse to the price of gold, and they have margins.
We would have $800 left to make other types of contracts.

File: README.md (hunk 2 above)
Comment (zack-bitcoin) [inline README.md]: I said something confusing things before. sorry.

of course the interest cost exists.
How many bitcoin would you pay right now to get 10 bitcoin in a month?
10 bitcoin minus that number is the monthly interest rate for bitcoin.

> If we can find some use to those bonded reserve-pool collateral tokens, we'll implement it.

That is the exact opposite of what I am suggesting. you are currently reusing the bonds for 2 purposes. I am telling you that you can't reuse bonds that way.

> I'm not convinced that channels are the solution

Neither am I. You don't need channels to solve this problem. That is just how I happened to solve it for flying fox. DPOS could solve this problem for example.

>Adding complexity here will likely skew the incentives and make the system insecure.
You are being complicated by reusing bonds. I am trying to convince you to simplify it. use bonds for only security against double-spend. Don't reuse them for deciding who has validating power.

If you reuse bonds for both protecting against double-signing, and for deciding who has validating power, it will do both things poorly.
It will be too many bonds for stopping double-signing, so you will waste money by the interest rate.
It will be too few bonds for deciding who should have consensus, so you will have retirement attacks.

File: README.md (hunk 3 above)
Comment (jaekwon) [inline README.md]: Are you assuming some oracle that determines the price of the underlying asset for settling this contract in the case of dispute?

File: README.md (hunk 2 above)
Comment (jaekwon) [inline README.md]: > It will be too many bonds for stopping double-signing, so you will waste money by the interest rate.

Assume 0 in the reserve-pool.  There's no money wasted, 100% of the gnuts represent claim to fees, which arrises from the network.  Gnuts != money.  Gnuts may have high value but they can't be spent or used anywhere else, so there's no opportunity cost.

The reserve-pool tokens arguably incur opportunity cost for being bonded, but they don't decide who the validators are.  Add to the reserve-pool until marginal cost == marginal return.  There aren't "too many bonds".

> It will be too few bonds for deciding who should have consensus, so you will have retirement attacks.

A significant coalition can try to attack the system, but they will get slashed.  Those who remain will own more relative gnuts.  The system is antifragile.  But this is assuming that there will even be profitable retirement attacks.  Anyone can throttle their txs to limit exposure.

File: README.md (hunk 3 above)
Comment (zack-bitcoin) [inline README.md]: yes.

You don't need channels to fix this bug. Channels are how I happened to solve it for flying fox.

File: README.md (hunk 2 above)
Comment (zack-bitcoin) [inline README.md]: The inability to spend gnuts elsewhere does not change the fact that you are losing value by the interest rate. For example, if I buy gold, my money becomes less liquid. By holding gold I am forgoing the possibility to take advantage of a sudden change in prices, like a sale on pinto beans.
Similarly, if I am holding gnuts, I am forgoing the ability to take advantage of a sale. When value is inside of a less-liquid location, the owner is losing value by the interest rate. If you don't pay to cover his cost, he wont invest.

If we can't spend from the reserve pool, then the money is already deleted. There is no cost to draining it for a retirement attacker.
The cost of refilling an empty pool is carried by the honest validators who remain.

If control over the reserve pool is proportionate to how many Gnuts you own, then we need to examine the combined ownership of both cryptoeconomically.
If this combo-asset is doing both things as once, then it must be doing them both poorly.
Too many bonds for stopping double-singing, but too few to stop retirement attacks.

just because you balance marginal cost and marginal return doesn't mean it is a cryptoeconomically sound design. Reusing the same asset for these 2 different goals is a bad idea. Each goal puts different restrictions on the bonds. The combination of restrictions is much worse than either on it's own.

If an amount of value equal to 1/2 (the value of the reserve pool + the total value of the gnuts) should ever flow through a single block, it will be profitable for a coalition to double-sign. This is a retirement attack. They don't care about getting slashed, because they are stealing more than they can lose.

If you make the reserve pool bigger, you will increase security against retirement attacks. If you make the pool smaller, the blockchain will be more affordable to run.

Using the same bonds for both these things makes it so you can't optimize both, you are finding a semi-bad medium between 2 bad alternatives.

Comment (zack-bitcoin) [inline README.md]: I wrote a doc trying to clarify this idea: https://github.com/BumblebeeBat/FlyingFox/blob/master/docs/2_types_of_bonds.md

Comment (jaekwon) [inline README.md]: How is putting it in a channel going to help, if you can spend it?  Why would anyone want to get in a channel with you when you might get the money slashed?

>  When value is inside of a less-liquid location, the owner is losing value by the interest rate.

Eh, except when the less-liquid thing is earning you revenue.  We're arguing about definitions here, I don't call that "losing value by interest rate".  It's about opportunity cost, and the price of gnuts will reflect that.  There's a huge difference between parking US dollars as collateral, vs holding gnuts.  The former clearly incurs costs due to the inflation of the dollar and opportunity costs, the latter doesn't.  There's nothing better for you to do with your gnuts than stake them.  You can sell them and do something else with your money, but if you're a good validator and the market is efficient, you won't be able to do better than keep your gnuts staked.

I'm pretty sure the confusion is due to you assuming that gnuts are like currency.  It is not.

> If we can't spend from the reserve pool, then the money is already deleted.

Yes, there should be a mechanism to get the money out of the reserve pool.

> Too many bonds for stopping double-singing, but too few to stop retirement attacks.

I don't understand why you're making this distinction.  The point of preventing double-signing is to prevent retirement attacks.

> If an amount of value equal to 1/2 (the value of the reserve pool + the total value of the gnuts) should ever flow through a single block, it will be profitable for a coalition to double-sign.

Well, yeah.  Don't do that.  Anyone can figure out how to transfer funds over time to prevent such a thing.  The amount at stake determines how much volume is safe to transfer, and thus more stake == more volume, and more fees.

Comment (zack-bitcoin) [inline README.md]: For example, if I offer to give you a $10 reward for either lending me $100 or $1000 for 24 hours, which would you prefer to lend me?
Both investments are profitable. 

Similarly, locking up a large portion of the funds all the time is more expensive than only locking up a smaller portion.

"retirement attack" was an unfortunate use of words. I rewrote the document https://github.com/BumblebeeBat/FlyingFox/blob/master/docs/2_types_of_bonds.md

Comment (jaekwon) [inline README.md]: You keep reducing the analogy to money, but the point is that gnuts aren't money.  Say you're a captain and you have a choice of loaning your money for interest, or purchasing a ship.  Most people can't make use of a ship but you're an expert at it.  Most people will incur costs by hodling a ship, even if it has resell value.  But the value of a ship to you is greater than that, because you can charge people for the transport of goods.  You're not locking up your money in a ship, you bought it to use it, and later you can sell it to other captains.

Comment (jaekwon) [inline README.md]: In GnuClear, in order for a double spend to succeed, 1/3+ of all the ships will explode.  Strange kind of ship, but there it is.

Comment (zack-bitcoin) [inline README.md]: yes, it does sound like GnuClear will have the correct amount of money held in bonds ready to be destroyed. That is one important quality for security, and it indeed protects from consensus forks. 

We also need to worry about soft forks.

An attacker could be interested in destroying as much money in the chain as possible. They could be willing to destroy one of their own money in exchange to destroy 2 moneys on the chain.

How much on-chain value can I destroy in exchange for sacrificing all the ships?
If the on-chain value destroyed is more than 2x as much as all the ships, I am willing to make the trade.

Since the validators can execute any soft fork, they should be able to destroy a lot of value.

I rewrote it again https://github.com/BumblebeeBat/FlyingFox/blob/master/docs/2_types_of_bonds.md
